numba/tests/test_np_functions.py

```diff
@@ -1327,6 +1327,67 @@ def nonconst_side(a, v, side='left'):
         )
         check(a, v)
 
+    def test_searchsorted_supplemental(self):
+        pyfunc = searchsorted
+        cfunc = jit(nopython=True)(pyfunc)
+
+        pyfunc_left = searchsorted_left
+        cfunc_left = jit(nopython=True)(pyfunc_left)
+
+        pyfunc_right = searchsorted_right
+        cfunc_right = jit(nopython=True)(pyfunc_right)
+
+        def check(a, v):
+            expected = pyfunc(a, v)
+            got = cfunc(a, v)
+            self.assertPreciseEqual(expected, got)
+
+            expected = pyfunc_left(a, v)
+            got = cfunc_left(a, v)
+            self.assertPreciseEqual(expected, got)
+
+            expected = pyfunc_right(a, v)
+            got = cfunc_right(a, v)
+            self.assertPreciseEqual(expected, got)
+
+        element_pool = list(range(-5, 50))
+        element_pool += [np.nan] * 5 + [np.inf] * 3 + [-np.inf] * 3
+
+        for i in range(1000):
+            sample_size = self.rnd.choice([5, 10, 25])
+
+            # `a` and `v` not sorted; either may have repeating values
+            a = self.rnd.choice(element_pool, sample_size)
+            v = self.rnd.choice(element_pool, sample_size + (i % 3 - 1))
+
+            # output should match numpy regardless of whether `a` is sorted
+            check(a, v)
+            check(np.sort(a), v)
+
+        ones = np.ones(5)
+        nans = np.full(len(ones), fill_value=np.nan)
+        check(ones, ones)
+
+        # `a` and / or `v` full of nans
+        check(ones, nans)
+        check(nans, ones)
+        check(nans, nans)
+
+        # `v` is zero size
+        a = np.arange(1)
+        v = np.arange(0)
+        check(a, v)
+
+        # `a` and `v` booleans
+        a = np.array([False, False, True, True])
+        v = np.array([False, True])
+        check(a, v)
+
+        # `v` is a (scalar) boolean
+        a = [1, 2, 3]
+        v = True
+        check(a, v)
+
```

Comment:
Could you also include a test for complex types? The following is crashing on your PR:

```python
import numpy as np
from numba import njit

side = 'left'

@njit
def foo(a, v):
    return np.searchsorted(a, v, side='left')

a = np.zeros(9, dtype=np.complex64)
a.real += [0, 0, 1, 1, 0, 1, np.nan, np.nan, np.nan]
a.imag += [0, 1, 0, 1, np.nan, np.nan, 0, 1, np.nan]
v = a
print()
print(a, v, side)
print(foo(a, v))
print(a.searchsorted(v, side=side))
```

Reply:
Ah yes, I think that snippet will similarly raise in the current implementation / main - in that regard, it's not a behaviour which is changed or introduced in this PR.

I have a feeling that numba in general doesn't support the ordering of complex numbers - but I guess it would be possible to branch this implementation to do whatever numpy does (which would make it an exception, I _think_).

I guess the following could also be caught handled.

```python
import numpy as np
from numba import njit

@njit
def foo(a, v):
    return np.searchsorted(a, v)

a = np.array(['1', '2', '3'])
v = np.array(['2', '4'])

print(foo(a, v))
```

I'll have a look at these things when I get some time.

Reply:
@guilhermeleobas in the last few commits, I have extended support for complex dtypes and handled the example in the prior comment.

As this involved implementing a lex sort for complex numbers a la numpy, I plumbed it into `np.sort` (and `array.sort`) too since it was a very small incremental change. If this conflation of changes seems unwise, let me know and I will extract it into a different PR.

There are some residual edge cases not handled but they never have been and would require a lot more work, so I'd propose to cut the scope here and would kindly request a re-review when you have a mo.

Reply:
Thanks @rjenc29. Could you create an issue to track this limitation?